Use Standalone Python, rather than a BeeWare-specific Python build

freakboy3742 opened the pull request:
Indygreg's [Python Standalone Builds](https://github.com/indygreg/python-build-standalone) is a project that provides tarballs of Python releases, similar to BeeWare's [Python-Linux-support](https://github.com/beeware/Python-Linux-support); however, they're being more actively maintained and audited for external dependencies.

Both AppImage and Flatpak can use these builds, providing a number of benefits:

1. One less project for BeeWare to maintain on it's own
2. Flatpak builds are significantly faster, as they don't need to start with a compile of CPython (3m vs 11m)
3. AppImage builds are significantly faster, as the Docker image used to build the AppImage doesn't need to add the Deadsnakes repos and download Python (and it's dependencies) (1.5m vs 3m). Instead, the AppImage can use the standalone build to provide pip and python.
4. AppImage builds are smaller, as there is less leakage of unintended dependencies.
5. AppImage and Flatpak now use a common execution base. 

This is work that came out of my initial experiments with Deb packaging; I figured I should wrap it up and submit it as an independent PR; even though AppImage is less significant now that System packaging is the default, having faster builds will be beneficial to CI. Reducing the maintenance burden of the Linux support package is also a nice benefit.

This will initially fail app verification checks on CI, because it requires template updates. 

Refs beeware/briefcase-template#61
Refs beeware/briefcase-linux-appimage-template#29
Refs beeware/briefcase-linux-flatpak-template#17

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

rmartin16 reviewed (commented):
Happy to see this; things went well for me building AppImages and Flatpaks.

Since `ARG PY_VERSION` is no longer being used in the AppImage Dockerfile and it isn't used in the Linux system Dockerfile, I think it'd be safe to remove it from the Docker image build if you want.

On a sidenote, I experimented with the additional versions of `x86_64` builds from Standalone Python that leverage modern x86 instructions. They worked fine with Flatpak and non-Docker AppImage builds. For Docker builds, I had to upgrade `gcc` to at least version 11 to compile `pycairo` but that wasn't too bad in the Dockerfile.

rmartin16 commented:
Another thing I realized....this should make it significantly easier to support different base images now (#947). The previous support package had a lot of more dependencies (for instance, `linuxdeploy` gives up after not finding `libncurses.so.5` in `ubuntu:20.04`).

Now, when I update Dockerfile to use `ubuntu:20.04`, the AppImage build finishes _and_ works.

I looked at using the `manylinux_2_28_x86_64` image but it's based on AlmaLinux 8; so, the Dockerfile will need to be rewritten.

freakboy3742 commented on src/briefcase/platforms/linux/__init__.py:
```diff
@@ -56,6 +56,22 @@ def parse_freedesktop_os_release(content):
+        the template, so this method won't be invoked
+        """
+        version, datestamp = support_revision.split("+")
+        return (
+            "https://github.com/indygreg/python-build-standalone/releases/download/"
+            f"{datestamp}/cpython-{support_revision}-x86_64-unknown-linux-gnu-install_only.tar.gz"
+        )
```
An oversight on my part - definitely agreed this shouldn't hardcode x86_64.

freakboy3742 commented:
> Another thing I realized....this should make it significantly easier to support different base images now (#947). 

Yeah - you're right - and it's not that hard to accomodate either.

I've pushed updates to allow for that, as well as a change to the briefcase-template that moves to manylinux2014 as a default base image (as the "oldest still supported manylinux"). 

The minor catch: Toga AppImages build, but don't run at present - it's not pulling in a GDK dependency for some reason.

freakboy3742 commented:
> Since `ARG PY_VERSION` is no longer being used in the AppImage Dockerfile and it isn't used in the Linux system Dockerfile, I think it'd be safe to remove it from the Docker image build if you want.

Agreed - I've removed it.

> On a sidenote, I experimented with the additional versions of `x86_64` builds from Standalone Python that leverage modern x86 instructions. They worked fine with Flatpak and non-Docker AppImage builds. For Docker builds, I had to upgrade `gcc` to at least version 11 to compile `pycairo` but that wasn't too bad in the Dockerfile.

Yeah - I know they exist, but I'm not sure how we'd be able control or opt-into their use in a way that makes sense. The naïve approach would be to allow the arch tag to be user-customisable; but then how do we ensure that the build machine or docker image support those instructions?

rmartin16 reviewed (changes requested):
Code review and initial testing comments.

> The minor catch: Toga AppImages build, but don't run at present - it's not pulling in a GDK dependency for some reason.

I tracked this down I think.

When I look at AppImage builds on `ubuntu:18.04`, `libgtk-3.so` ends up getting pulled in by a "printbackend".
```
Docker| [gtk/stdout] Deploying dependencies for ELF file /app/gtkubuntu.AppDir/usr/lib/x86_64-linux-gnu/gtk-3.0/3.0.0/printbackends/libprintbackend-cups.so                                                                       
Docker| [gtk/stdout] Deploying shared library /usr/lib/x86_64-linux-gnu/libgtk-3.so.0 
```

However, on `quay.io/pypa/manylinux2014_x86_64`, the "printbackends" aren't even being considered by `linuxdeploy` for inclusion in the AppImage.

In the gtk plugin, it [copies](https://github.com/linuxdeploy/linuxdeploy-plugin-gtk/blob/f46da605d876b4eb05cf91092a830804daf7ec3f/linuxdeploy-plugin-gtk.sh#L233) the GTK3 modules from `/usr/lib64/gtk-3.0` in to the base of the AppDir; therefore, all these files and directories end up in `/usr/lib64` within the AppDir.

However, `linuxdeploy` does not iterate the `/usr/lib64` directory; it [iterates](https://github.com/linuxdeploy/linuxdeploy/blob/dee359b179844e5d0dbe449e559a7dc3c402a32b/src/core/appdir.cpp#L812) `/usr/lib` inside AppDir.

Therefore, the AppImage build doesn't pull in all the needed dependencies that were previously being included by `linuxdeploy` reading the dependencies from these modules.

When I copy the GTK3 module directories from `/usr/lib64` in to `/usr/lib` in AppDir and rerun the AppImage build, I am able to run the HelloWorld app (with only installing `libthai-dev libfribidi0 libfontconfig1 libx11-6 libegl1 libgl1 libharfbuzz0b` and running `update-mime-database /usr/share/mime`).

I can only imagine that the intention of the gtk plugin copying the GTK modules in to AppDir....is so they and their dependencies are included in the AppImage. If you agree, I can log this issue with linuxdeploy/linuxdeploy-plugin-gtk.

rmartin16 commented on docs/reference/platforms/linux/appimage.rst:
```diff
@@ -16,32 +16,13 @@ library versions present on each distribution. An AppImage can be executed on
-   version of libc that is available on the distribution you use. If you build
-   using Ubuntu 19.10, for example, you can expect that only people on the most
-   recent versions of another distribution will be able to run your AppImage.
+To ensure that an application is built in an environment that is as compatible
+as possible, Briefcase builds AppImages inside Docker. The Docker base image is
+ubuntu:18.04 by default, but can be configured to any `manylinux
+<https://github.com/pypa/manylinux>`__ base using the `manylinux` application
+configuration option. While it is possible to build AppImages *without* Docker,
+it is highly recommended that you do not, as the resulting AppImages will not be
+as portable as they could otherwise be.
```
re-reading this now....i think it may actually be clear that ubuntu is the default when `manylinux` isn't set....but `briefcase new` will set it. I guess I was just thrown off by `ubuntu:18.04 by default` when that's conditional on an overriding default....

freakboy3742 replied:
Agreed it's ambiguous. I'll tweak the language.

rmartin16 commented on src/briefcase/platforms/linux/appimage.py:
```diff
@@ -120,21 +125,48 @@ def verify_host(self):
-    def support_package_filename(self, support_revision):
-        """The query arguments to use in a support package query request."""
-        return f"Python-{self.python_version_tag}-linux-{self.tools.host_arch}-support.b{support_revision}.tar.gz"
+    def finalize_app_config(self, app: AppConfig):
```
I also just noticed that this doesn't run if the `create` command is only run implicitly; for instance, as part of `briefcase build`. This seems to be a consequence of `if hasattr(app, "__draft__")` back up in `base.py`.

freakboy3742 replied:
Ah - of course - I need to move the finalise to the mixin location, rather than being defined only in create. Nice catch.

rmartin16 commented on src/briefcase/platforms/linux/appimage.py:
```diff
@@ -120,21 +125,48 @@ def verify_host(self):
+        context = super().output_format_template_context(app)
+        # Add the manylinux tag to the template context.
+        try:
+            context["manylinux_tag"] = f"{app.manylinux}_{self.tools.host_arch}"
```
> users will regularly see entire rebuilds of their Docker image.

Actually, I guess this would only be true if something re-pulled the `latest` tag....which is possible but won't just happen on its own. The reproducibility argument probably still stands, though, for environments like CI, for instance.

freakboy3742 replied:
I guess it can't hurt to expose the tag as a configurable option - I'll add that in.

freakboy3742 commented:
> In the gtk plugin, it [copies](https://github.com/linuxdeploy/linuxdeploy-plugin-gtk/blob/f46da605d876b4eb05cf91092a830804daf7ec3f/linuxdeploy-plugin-gtk.sh#L233) the GTK3 modules from `/usr/lib64/gtk-3.0` in to the base of the AppDir; therefore, all these files and directories end up in `/usr/lib64` within the AppDir.

Ugh. I was afraid of that. On the upside, we've got patches into the linuxdeploy GTK plugin in the past, so this is fixable. But... ugh.

> I can only imagine that the intention of the gtk plugin copying the GTK modules in to AppDir....is so they and their dependencies are included in the AppImage. If you agree, I can log this issue with linuxdeploy/linuxdeploy-plugin-gtk.

Definitely worth logging an issue - and the plugin is a shell script, so the fix is probably close to a one liner.

freakboy3742 commented on src/briefcase/platforms/linux/appimage.py:
```diff
@@ -120,21 +125,48 @@ def verify_host(self):
+    def finalize_app_config(self, app: AppConfig):
+        """If we're *not* using Docker, verify that we're actually on Linux."""
+        if not self.use_docker:
+            self.logger.info(
+                """\
+*************************************************************************
+** WARNING: Building a Local AppImage!                                 **
+*************************************************************************
+
+    You are building an AppImage outside Docker. The resulting AppImage
+    will work, but will not be as portable as a Docker-based AppImage.
+    Any `manylinux` setting will be ignored.
+
+*************************************************************************
+"""
```
Yup - that was a typo. Fixed.

freakboy3742 commented:
I've submitted linuxdeploy/linuxdeploy-plugin-gtk#43 as a fix for the GDK deployment issue, and update briefcase-template to include a couple more libraries that are needed at runtime. 

There's still an issue with (non-critical) libgvfsdbus, but that seems to be very deep in the weeks, and is at least partially hitting up against the fact that GTK plugins expect to be in specific directories, but linuxdeploy moves all libraries into the /usr/lib directory. You can fix the search directory by defining GIO_MODULE_DIR into the app hooks - but then the modules crash because they have RPATH references that haven't been updated.

rmartin16 commented:
After they merged https://github.com/linuxdeploy/linuxdeploy-plugin-gtk/pull/43, I gave it a try....but `libgtk-3.so` still isn't being pulled in to the AppImage for me. Is it working for you?

freakboy3742 commented:
It did - but I needed to manually delete the cached tools/linuxdeploy_plugins/gtk folder for it to work. Annoyingly, `briefcase upgrade linuxdeploy` didn't pick up the update... so there's a bug in that logic.

rmartin16 commented:
Interesting....Ive deleted my entire briefcase cache directory and verified the changes from that PR are in my copy of the plugin:
```
Docker| [gtk/stderr] ++ find /usr/lib64 '(' -type l -o -type f ')' -name 'libgdk-*.so*' -print0
```
Yet, I don't get `libgtk-3.so` in the AppImage:
```
❯ ./Hello_World-0.0.1-x86_64.AppImage --appimage-extract | grep libgtk*

❯ 
```
I am definitely interested in knowing what I might be missing....I'd ask you to share the produced AppImage but that may be difficult to upload somewhere....maybe your Briefcase log?

freakboy3742 commented:
Well this is spicy... I just rebuilt my AppImage so I could provide a log... and this time it *didn't* work - I'm getting a completely new error.

Previously (with the old plugin script), I was getting this error when GDK wasn't found.
```
Failed to load shared library 'libgdk-3.so.0' referenced by the typelib: /lib64/libpangoft2-1.0.so.0: undefined symbol: g_atomic_rc_box_acquire                                                 
...
File "/tmp/.mount_Toga_TgWuALN/usr/app_packages/gi/overrides/__init__.py", line 195, in override
    assert g_type != TYPE_NONE
AssertionError 
```

My first build once I cleared the cached file succeeded, and the run worked. I just deleted the build directory and re-ran; now, I'm getting:
```
Failed to load shared library 'libgtk-3.so.0' referenced by the typelib: /lib64/libgtk-3.so.0: undefined symbol: gdk_broadway_window_get_type 
...
File "/tmp/.mount_Toga_TgjqpIb/usr/app_packages/gi/types.py", line 247, in __init__
    register_interface_info(cls.__info__.get_g_type())
TypeError: must be an interface  
```

FWIW, build log for the second attempt:
[briefcase.2023_03_13-07_01_09.build.log](https://github.com/beeware/briefcase/files/10952342/briefcase.2023_03_13-07_01_09.build.log)

The build directory is showing libgdk-3.0.so as being deployed. Not sure where the missing symbol *should* be. 

FWIW, I'm building in Docker on macOS, and running on a Fedora 36 parallels instance.

rmartin16 commented:
I think this is where I've been a little confused.....you updated the plugin script to pull in `libgdk-3` libraries....but we're missing `libgtk-3` libraries. Should the `libgdk` libraries be pulling `libgtk` libraries somehow?

freakboy3742 commented:
Oh! Clearly I need the coffee to kick in... I misread that as another g**d**k error. 

So... I have no idea why gtk isn't being picked up - that *should* be the entire modus operandi of the script. I'm guessing it worked previously because it wasn't a clean build folder and libgtk was still there from a historical pass. Not sure what I've done to break that, though...

rmartin16 commented:
Ok. It seems we were hitting different errors before the plugin was updated. I was getting an error on `libgtk-3.so` missing from the beginning for some reason.

Nonetheless, we're missing it because nothing that `linuxdeploy` is scanning depends on it (...obviously); we were only getting it pulled in on `ubuntu:18.04` because the `linuxdeploy` was scanning the GTK modules.....but that won't happen on RHEL derived images because of `/usr/lib64`.

Looking at the CI test case for fedora in the plugin repo, they are testing against a executable that dynamically links `libgtk.so`.....hence, GTK is getting pulled in.

freakboy3742 commented:
> Ok. It seems we were hitting different errors before the plugin was updated. I was getting an error on `libgtk-3.so` missing from the beginning for some reason.
> 
> Nonetheless, we're missing it because nothing that `linuxdeploy` is scanning depends on it (...obviously); we were only getting it pulled in on `ubuntu:18.04` because the `linuxdeploy` was scanning the GTK modules.....but that won't happen on RHEL derived images because of `/usr/lib64`.
> 
> Looking at the CI test case for fedora in the plugin repo, they are testing against a executable that dynamically links `libgtk-3.so`.....hence, GTK is getting pulled in.

That all makes sense - what doesn't make sense is why it has *ever* worked for me... I have done plenty of clean builds over the last couple of days, so how have I *ever* got gtk linked in successfully?

The immediate fix is to be explicit about libgtk - adding the following to the array of copied libraries:
```
    "$gtk_libdir"     "libgtk-$DEPLOY_GTK_VERSION.so*"
```

however, I wonder if there's a better approach that involves fixing the path that is copying modules so that the AppImage isn't using both lib and lib64.

rmartin16 commented:
> > Ok. It seems we were hitting different errors before the plugin was updated. I was getting an error on `libgtk-3.so` missing from the beginning for some reason.
> > Nonetheless, we're missing it because nothing that `linuxdeploy` is scanning depends on it (...obviously); we were only getting it pulled in on `ubuntu:18.04` because the `linuxdeploy` was scanning the GTK modules.....but that won't happen on RHEL derived images because of `/usr/lib64`.
> > Looking at the CI test case for fedora in the plugin repo, they are testing against a executable that dynamically links `libgtk-3.so`.....hence, GTK is getting pulled in.
> 
> That all makes sense - what doesn't make sense is why it has _ever_ worked for me... I have done plenty of clean builds over the last couple of days, so how have I _ever_ got gtk linked in successfully?

I think I figured this one out. I noticed during a rebuild of the AppImage that the Docker command that Briefcase ran was more than I was expecting.

Fresh build of the AppImage:
```
docker run --rm 
	--volume /home/russell/tmp/beeware/helloworld/build/helloworld/linux/appimage:/app:z 
	--volume /home/russell/.cache/briefcase:/home/brutus/.cache/briefcase:z 
	--env DEBUG=1 
	--env DEPLOY_GTK_VERSION=3 
	--env 'PATH=/home/brutus/.cache/briefcase/tools/linuxdeploy_plugins/gtk:/app/Hello World.AppDir/usr/python/bin:/opt/rh/devtoolset-10/root/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' 
	--env VERSION=0.0.1 
	--env DISABLE_COPYRIGHT_FILES_DEPLOYMENT=1 
	--env APPIMAGE_EXTRACT_AND_RUN=1 
	--env ARCH=x86_64 
	--workdir /app 
	briefcase/com.example.helloworld:manylinux2014-appimage
	/home/brutus/.cache/briefcase/tools/linuxdeploy-x86_64.AppImage 
		--appdir '/app/Hello World.AppDir' 
		--desktop-file '/app/Hello World.AppDir/com.example.helloworld.desktop' 
		--output appimage 
		--deploy-deps-only '/app/Hello World.AppDir/usr/app_packages/cairo'
		--deploy-deps-only '/app/Hello World.AppDir/usr/app_packages/gi' 
		--deploy-deps-only '/app/Hello World.AppDir/usr/python/lib'
		--plugin gtk
```

Rebuild of the same AppImage:
```
docker run --rm 
	--volume /home/russell/tmp/beeware/helloworld/build/helloworld/linux/appimage:/app:z 
	--volume /home/russell/.cache/briefcase:/home/brutus/.cache/briefcase:z 
	--env DEBUG=1 
	--env DEPLOY_GTK_VERSION=3 
	--env 'PATH=/home/brutus/.cache/briefcase/tools/linuxdeploy_plugins/gtk:/app/Hello World.AppDir/usr/python/bin:/opt/rh/devtoolset-10/root/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' 
	--env VERSION=0.0.1 
	--env DISABLE_COPYRIGHT_FILES_DEPLOYMENT=1 
	--env APPIMAGE_EXTRACT_AND_RUN=1 
	--env ARCH=x86_64 
	--workdir /app 
	briefcase/com.example.helloworld:manylinux2014-appimage 
	/home/brutus/.cache/briefcase/tools/linuxdeploy-x86_64.AppImage 
		--appdir '/app/Hello World.AppDir' 
		--desktop-file '/app/Hello World.AppDir/com.example.helloworld.desktop' 
		--output appimage 
		--deploy-deps-only '/app/Hello World.AppDir/usr/app_packages/cairo' 
		--deploy-deps-only '/app/Hello World.AppDir/usr/app_packages/gi' 
		--deploy-deps-only '/app/Hello World.AppDir/usr/lib' 
		--deploy-deps-only '/app/Hello World.AppDir/usr/lib64/gdk-pixbuf-2.0/2.10.0/loaders' 
		--deploy-deps-only '/app/Hello World.AppDir/usr/lib64/gtk-3.0/3.0.0/printbackends' 
		--deploy-deps-only '/app/Hello World.AppDir/usr/lib64/gtk-3.0/modules' 
		--deploy-deps-only '/app/Hello World.AppDir/usr/python/lib'
		--plugin gtk
```

So, on subsequent rebuilds of the AppImage, Briefcase picks up the GTK libraries that were copied in to `/usr/lib64` inside AppDir on the previous run and explicitly tells `linuxdeploy` to find their dependencies and include them....which will include `libgtk-3.so`. I was always running `rm -rf dist build` before rebuilds.

rmartin16 commented:
After discussion in #AppImage IRC, I created https://github.com/linuxdeploy/linuxdeploy-plugin-gtk/issues/44; although, I'm mostly expecting one of us will need to take action for it to resolve soonish. I'll let you know if/when I'm able to make any progress.

freakboy3742 commented:
Ok - [here's a patched linuxdeploy-plugin-gtk branch](https://github.com/freakboy3742/linuxdeploy-plugin-gtk/tree/lib64-fix) that work in my testing. It patches any reference to `/usr/lib64` (or any other possible lib dir) into `$AppDir/usr/lib` 

There's still an open issue where it fails to find GVFS/DBUS stuff at startup; this logs a message on *some* platforms, but doesn't cause any runtime problems that I can see.

@rmartin16 Can you confirm if this works for you?

rmartin16 commented:
> Ok - [here's a patched linuxdeploy-plugin-gtk branch](https://github.com/freakboy3742/linuxdeploy-plugin-gtk/tree/lib64-fix) that work in my testing. It patches any reference to `/usr/lib64` (or any other possible lib dir) into `$AppDir/usr/lib`
> 
> There's still an open issue where it fails to find GVFS/DBUS stuff at startup; this logs a message on _some_ platforms, but doesn't cause any runtime problems that I can see.
> 
> @rmartin16 Can you confirm if this works for you?

Got some really funky directories being created out of the gate....found the expansion replacement was missing `$LD_GTK_LIBRARY_PATH` for the `mkdir`. My suggested changes are below.

```diff
diff --git a/linuxdeploy-plugin-gtk.sh b/linuxdeploy-plugin-gtk.sh
index d49fe18..0730b5b 100755
--- a/linuxdeploy-plugin-gtk.sh
+++ b/linuxdeploy-plugin-gtk.sh
@@ -70,8 +70,8 @@ copy_lib_tree() {
     local dst="${*:$#}"
 
     for elem in "${src[@]}"; do
-        mkdir -p "${dst::-1}${elem///usr/lib}"
-        pushd $LD_GTK_LIBRARY_PATH
+        mkdir -p "${dst::-1}${elem/$LD_GTK_LIBRARY_PATH//usr/lib}"
+        pushd "$LD_GTK_LIBRARY_PATH"
         cp "$(realpath --relative-to="$LD_GTK_LIBRARY_PATH" "$elem")" --archive --parents --target-directory="$dst/usr/lib" $verbose
         popd
     done
@@ -273,7 +273,7 @@ EOF
         copy_lib_tree "$gtk4_libdir" "$APPDIR/"
         cat >> "$HOOKFILE" <<EOF
 export GTK_EXE_PREFIX="\$APPDIR/$gtk4_exec_prefix"
-export GTK_PATH="\$APPDIR/$gtk4_path"
+export GTK_PATH="\$APPDIR/${gtk4_path/$LD_GTK_LIBRARY_PATH//usr/lib}}"
 EOF
         ;;
     *)
```

The `libcanberra` library is still having issues.....`libcanberra-gtk-module.so` cannot find `libcanberra-gtk3.so.0` which is in `usr/lib/` of the AppImage.....but that is not part of its runpath.

<details>
<summary>LD_DEBUG Details</summary>

```
      1235:	file=libcanberra-gtk3.so.0 [0];  needed by /tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/libcanberra-gtk-module.so [0]
      1235:	find library=libcanberra-gtk3.so.0 [0]; searching
      1235:	 search path=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/glibc-hwcaps/x86-64-v3:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/glibc-hwcaps/x86-64-v2:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls/x86_64/x86_64:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls/x86_64:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls/x86_64:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/x86_64/x86_64:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/x86_64:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/x86_64:/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules		(RUNPATH from file /tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/libcanberra-gtk-module.so)
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/glibc-hwcaps/x86-64-v3/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/glibc-hwcaps/x86-64-v2/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/tls/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/tmp/appimage_extracted_d72ef1d62d1207d4721ba5c86c0637df//usr/lib/gtk-3.0/modules/libcanberra-gtk3.so.0
      1235:	 search cache=/etc/ld.so.cache
      1235:	 search path=/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v3:/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v2:/lib/x86_64-linux-gnu/tls/x86_64/x86_64:/lib/x86_64-linux-gnu/tls/x86_64:/lib/x86_64-linux-gnu/tls/x86_64:/lib/x86_64-linux-gnu/tls:/lib/x86_64-linux-gnu/x86_64/x86_64:/lib/x86_64-linux-gnu/x86_64:/lib/x86_64-linux-gnu/x86_64:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v3:/usr/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v2:/usr/lib/x86_64-linux-gnu/tls/x86_64/x86_64:/usr/lib/x86_64-linux-gnu/tls/x86_64:/usr/lib/x86_64-linux-gnu/tls/x86_64:/usr/lib/x86_64-linux-gnu/tls:/usr/lib/x86_64-linux-gnu/x86_64/x86_64:/usr/lib/x86_64-linux-gnu/x86_64:/usr/lib/x86_64-linux-gnu/x86_64:/usr/lib/x86_64-linux-gnu:/lib/glibc-hwcaps/x86-64-v3:/lib/glibc-hwcaps/x86-64-v2:/lib/tls/x86_64/x86_64:/lib/tls/x86_64:/lib/tls/x86_64:/lib/tls:/lib/x86_64/x86_64:/lib/x86_64:/lib/x86_64:/lib:/usr/lib/glibc-hwcaps/x86-64-v3:/usr/lib/glibc-hwcaps/x86-64-v2:/usr/lib/tls/x86_64/x86_64:/usr/lib/tls/x86_64:/usr/lib/tls/x86_64:/usr/lib/tls:/usr/lib/x86_64/x86_64:/usr/lib/x86_64:/usr/lib/x86_64:/usr/lib		(system search path)
      1235:	  trying file=/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v3/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v2/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/tls/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/tls/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64-linux-gnu/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v3/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/glibc-hwcaps/x86-64-v2/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/tls/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/tls/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64-linux-gnu/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/glibc-hwcaps/x86-64-v3/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/glibc-hwcaps/x86-64-v2/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/tls/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/tls/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/lib/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/glibc-hwcaps/x86-64-v3/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/glibc-hwcaps/x86-64-v2/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/tls/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/tls/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/tls/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/x86_64/libcanberra-gtk3.so.0
      1235:	  trying file=/usr/lib/libcanberra-gtk3.so.0
Gtk-Message: 18:11:39.756: Failed to load module "canberra-gtk-module"
```
</details>

I'll probably look at this more a little later or tomorrow.

rmartin16 commented:
> The `libcanberra` library is still having issues.....`libcanberra-gtk-module.so` cannot find `libcanberra-gtk3.so.0` which is in `usr/lib/` of the AppImage.....but that is not part of its runpath.

The "best" solution I have for this right now is adding `LD_LIBRARY_PATH="${APPDIR}/usr/lib:${LD_LIBRARY_PATH}"` before the call to `${APPDIR}/usr/python/bin/python3`....I put the question in #AppImage IRC as well...

rmartin16 commented:
Another thing....

The new Python binary has a dependency on `libcrypt.so.1`, among others, but this one should be sourced from within the AppImage. Unfortunately, since we don't tell `linuxdeploy` about this Python binary, it does not update it to use `$ORIGIN/../../lib` as a `RUNPATH`.

I'm not sure the best way to tell `linuxdeploy` about `usr/python/bin` though....only the app template knows about it....maybe a `linuxdeploy_include_deps` option in `briefcase.toml`...?

(the previously mentioned `LD_LIBRARY_PATH` setting would avoid this problem....but I do not think that's a good solution to any of this)

freakboy3742 commented:
> Got some really funky directories being created out of the gate....found the expansion replacement was missing `$LD_GTK_LIBRARY_PATH` for the `mkdir`. My suggested changes are below.

The `mkdir` correction looks like I forgot one last save before pushing 🤦. The quoting of the `pushd` and the GTK4 path fix are both good catches, though - I definitely missed those.
 
> The `libcanberra` library is still having issues.....`libcanberra-gtk-module.so` cannot find `libcanberra-gtk3.so.0` which is in `usr/lib/` of the AppImage.....but that is not part of its runpath.

Hrm... I wonder if there might be some additional module handling in the block at the end of the file...

rmartin16 commented:
> > The `libcanberra` library is still having issues.....`libcanberra-gtk-module.so` cannot find `libcanberra-gtk3.so.0` which is in `usr/lib/` of the AppImage.....but that is not part of its runpath.
> 
> Hrm... I wonder if there might be some additional module handling in the block at the end of the file...

While it says "manually setting rpath for GTK modules"....it's creating a bunch of symlinks...I don't think that's gonna help `/usr/lib/gtk-3.0/modules/libcanberra-gtk3-module.so` find `/usr/lib/libcanberra-gtk3.so.0`.....maybe if the symlinks went the other way and recreated `/usr/lib` within `/usr/lib/gtk-3.0/modules` for each module....

Nonetheless, that block should be updated with the rest of these changes:
```diff
diff --git a/linuxdeploy-plugin-gtk.sh b/linuxdeploy-plugin-gtk.sh
index d49fe18..ba3c91a 100755
--- a/linuxdeploy-plugin-gtk.sh
+++ b/linuxdeploy-plugin-gtk.sh
@@ -341,6 +342,6 @@ PATCH_ARRAY=(
 )
 for directory in "${PATCH_ARRAY[@]}"; do
     while IFS= read -r -d '' file; do
-        ln $verbose -sf "${file/\/usr\/lib\//}" "$APPDIR/usr/lib"
+        ln $verbose -sf "${file/$LD_GTK_LIBRARY_PATH\//}" "$APPDIR/usr/lib"
     done < <(find "$directory" -name '*.so' -print0)
 done
```

I haven't really grasped the motivation [here](https://github.com/linuxdeploy/linuxdeploy/blob/dee359b179844e5d0dbe449e559a7dc3c402a32b/src/core/appdir.cpp#L455) but libraries for the GTK modules just don't get stamped with a `RUNPATH` using a relative path from `$ORIGIN`....like `usr/python/lib/libpython3.so` does, for instance.

freakboy3742 commented:
> Nonetheless, that block should be updated with the rest of these changes:

Good catch; updated.

> I haven't really grasped the motivation [here](https://github.com/linuxdeploy/linuxdeploy/blob/dee359b179844e5d0dbe449e559a7dc3c402a32b/src/core/appdir.cpp#L455) but libraries for the GTK modules just don't get stamped with a `RUNPATH` using a relative path from `$ORIGIN`....like `usr/python/lib/libpython3.so` does, for instance.

Yeah - that whole setup is really weird. It seems like there's a huge conflict between GTK's desire to use folders for plugins and modules, and AppImages insistence that all .so objects are in `$APPDIR/usr/lib`, and it's not entirely clear to me what is meant to be fixed by what step in the process.

rmartin16 commented:
Also, since this is wonderful bash scripting and we're manually manipulating filepaths :)....you'll need to include the final slash...or perhaps add a replacement string of a period.
```diff
diff --git a/linuxdeploy-plugin-gtk.sh b/linuxdeploy-plugin-gtk.sh
index 61206fc..c506690 100755
--- a/linuxdeploy-plugin-gtk.sh
+++ b/linuxdeploy-plugin-gtk.sh
@@ -341,6 +341,6 @@ PATCH_ARRAY=(
 )
 for directory in "${PATCH_ARRAY[@]}"; do
     while IFS= read -r -d '' file; do
-        ln $verbose -sf "${file/$LD_GTK_LIBRARY_PATH/}" "$APPDIR/usr/lib"
+        ln $verbose -sf "${file/$LD_GTK_LIBRARY_PATH\//}" "$APPDIR/usr/lib"
     done < <(find "$directory" -name '*.so' -print0)
 done
```
These are the symlinks otherwise:
```
❯ ls build/helloworld/linux/appimage/Hello\ World.AppDir/usr/lib/
...
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-ani.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-ani.so
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-bmp.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-bmp.so
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-gif.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-gif.so
lrwxrwxrwx   54 russell 18 Mar 23:53  libpixbufloader-icns.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-icns.so
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-ico.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-ico.so
lrwxrwxrwx   56 russell 18 Mar 23:53  libpixbufloader-jasper.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-jasper.so
lrwxrwxrwx   54 russell 18 Mar 23:53  libpixbufloader-jpeg.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-jpeg.so
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-pnm.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-pnm.so
lrwxrwxrwx   54 russell 18 Mar 23:53  libpixbufloader-qtif.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-qtif.so
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-tga.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-tga.so
lrwxrwxrwx   54 russell 18 Mar 23:53  libpixbufloader-tiff.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-tiff.so
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-xbm.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xbm.so
lrwxrwxrwx   53 russell 18 Mar 23:53  libpixbufloader-xpm.so -> /gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-xpm.so
lrwxrwxrwx   58 russell 18 Mar 23:53  libprintbackend-cloudprint.so -> /gtk-3.0/3.0.0/printbackends/libprintbackend-cloudprint.so
lrwxrwxrwx   52 russell 18 Mar 23:53  libprintbackend-cups.so -> /gtk-3.0/3.0.0/printbackends/libprintbackend-cups.so
lrwxrwxrwx   52 russell 18 Mar 23:53  libprintbackend-file.so -> /gtk-3.0/3.0.0/printbackends/libprintbackend-file.so
lrwxrwxrwx   51 russell 18 Mar 23:53  libprintbackend-lpr.so -> /gtk-3.0/3.0.0/printbackends/libprintbackend-lpr.so
```

freakboy3742 commented:
I've [pushed the Linuxdeploy-plugin-gtk fixes upstream](linuxdeploy/linuxdeploy-plugin-gtk#45); if they're accepted, this PR should be ready for a final review and merge.

freakboy3742 commented:
@rmartin16 The Linuxdeploy-gtk-plugin patch has now been merged; care to take a final swing at a review?

rmartin16 commented:
All right....here's where we're at...

- New Python dependencies
  - Standalone Python depends on `libcrypt.so.1` - this is [understood](https://gregoryszorc.com/docs/python-build-standalone/main/running.html#runtime-requirements) and has been [problematic](https://gregoryszorc.com/docs/python-build-standalone/main/quirks.html#missing-libcrypt-so-1) in the past
  - We aren't telling `linuxdeploy` to update Standalone Python to use the `libcrypt` that's in the AppImage
  - [ ] So, we should make an explicit decision to tell `linuxdeploy` about Standalone Python
  - [ ] Or at least update the docs to say that produced AppImages will require `libcrypt` on host systems (since this behavior is not typical of producing an AppImage otherwise)

- Produced AppImages no longer support Ubuntu 18.04
  - toga requires `fribidi` on the host but the version of `fribidi` for Ubuntu 18.04 is too old and the version of `pango` from `manylinux2014` can't load it
  - `** (process:910): WARNING **: 23:27:42.351: Failed to load shared library 'libpango-1.0.so.0' referenced by the typelib: /tmp/appimage_extracted_3a06bfda9cd78e47fd6fa88549edf61c/usr/app_packages/gi/../../lib/libpango-1.0.so.0: undefined symbol: fribidi_get_par_embedding_levels_ex`
  - I guess I'm not terribly worried about this.....but I want to make sure it's known

- `linuxdeploy` and GTK modules are broken
  - Even with all the work to get the plugin to work with `/usr/lib64` and such....`linuxdeploy` doesn't update the GTK module libraries so they look in `/usr/lib` for the shared libraries they need
  - This seems to stem from `linuxdeploy` assuming a flat structure in `/usr/lib` instead of the nested directories that GTK uses for modules
  - The plugin already has a section at the very end that implies it's going to update the `RPATH` for module libraries to look in `/usr/lib`....but it doesn't actually do that...at all
  - [ ] The only workaround I can think of is setting `LD_LIBRARY_PATH` to `/usr/lib` in the startup script for the AppImage
  - [ ] All that said....this is basically status quo from where we were. so, if we doing nothing, we should probably take `libcanberra-gtk3` and `PackageKit-gtk3-module` outta `briefcase-template` since it doesn't do anything and will probably only serve to confuse people

- PySide6 Support
  - Support for PySide6 in AppImage was already pretty dubious....and I'm not sure the situation improves here or not...but I'll only really be interested in figuring this out if someone has a real use case....
  - That said, we should change a few things:
    - The `qt6-qtbase-gui` package isn't available in the distos used by `manylinux` images; we should just remove it
    - The PySide6 wheels are built on `manylinux_2_28`....so, we should default to using `manylinux_2_28` for it

- `musl` support
  - Standalone Python provides builds for `musl`....and there's `musllinux_1_1`
  - So, if anybody gets interested, it might be possible to add support for `musl` distros

I should be done testing, though; with these things addressed one way or the other, I should be good.

freakboy3742 commented:
> All right....here's where we're at...
> 
> * New Python dependencies
>     * []  Or at least update the docs to say that produced AppImages will require `libcrypt` on host systems (since this behavior is not typical of producing an AppImage otherwise)

Based on [the quirks discussion](https://gregoryszorc.com/docs/python-build-standalone/main/quirks.html#missing-libcrypt-so-1) I think I'm comfortable treating this as documentation, and saying "your Linux must be LSB-compliant, and RHEL isn't by default". I've added documentation to this effect.

> * Produced AppImages no longer support Ubuntu 18.04
>   * I guess I'm not terribly worried about this.....but I want to make sure it's known

Yeah - I'm in the same boat here. 18.04 is pretty old - as a user, you don't get to say "I want to use a 5 year old LSB, *and* have the bleeding edge of GUI packages.". I've documented this as a quirk.

> * `linuxdeploy` and GTK modules are broken
>     * [ ]  All that said....this is basically status quo from where we were. so, if we doing nothing, we should probably take `libcanberra-gtk3` and `PackageKit-gtk3-module` outta `briefcase-template` since it doesn't do anything and will probably only serve to confuse people

I'm seeing slightly better behavior when running on Fedora - if these packages are in the build environment, I don't get libcanberra and package kit errors at runtime. It's definitely not perfect, and it seems reasonable to assume that if the problems with linuxdeploy are ever fixed, these will be requirements, so I'm inclined to leave them in.

However, beyond that - I think "status quo, but now on manylinux" is close to the best case outcome. I'm pretty much at the end of my patience with AppImage generally. The meal you cook is only as good as the ingredients, and in this case, the ingredients... leave a bad taste in my mouth. AppImage is no longer the default packaging format, so our exposure to these bugs is smaller; I'm comfortable saying we can support AppImage, but our support is only as good as AppImage itself. I'll add a note to this effect.

> * PySide6 Support
>   
>   * Support for PySide6 in AppImage was already pretty dubious....and I'm not sure the situation improves here or not...but I'll only really be interested in figuring this out if someone has a real use case....
>   * That said, we should change a few things:
>     
>     * The `qt6-qtbase-gui` package isn't available in the distos used by `manylinux` images; we should just remove it

Removed.

>     * The PySide6 wheels are built on `manylinux_2_28`....so, we should default to using `manylinux_2_28` for it

I agree with the motivation... but if we don't have a configuration for PySide6, then I'm not sure this is worth adding the complexity to the base template. If someone wants to resolve PySide6 builds, then I'm sure this will be the first thing they need to change.

> * `musl` support
>   
>   * Standalone Python provides builds for `musl`....and there's `musllinux_1_1`
>   * So, if anybody gets interested, it might be possible to add support for `musl` distros

Agreed that this is something we could put on the feature list (as is providing more general customisation of the standalone image that is used, rather than harcoding the `linux-gnu-install_only` release. I can't say this is particularly high on my own personal priority list, especially given that we're directing attention away from AppImage and Flatpak. I'm mostly inclined to wait and see if someone actually requests musl support for AppImage, rather than encourage someone to add another configuration point that we'll have to maintain, but nobody actually uses.

> I should be done testing, though; with these things addressed one way or the other, I should be good.

👍 I've made the changes as described. Thanks for all the eyeballs on this - most appreciated.

rmartin16 commented:
> > * `linuxdeploy` and GTK modules are broken
> >   
> >   * [ ]   All that said....this is basically status quo from where we were. so, if we doing nothing, we should probably take `libcanberra-gtk3` and `PackageKit-gtk3-module` outta `briefcase-template` since it doesn't do anything and will probably only serve to confuse people
> 
> I'm seeing slightly better behavior when running on Fedora - if these packages are in the build environment, I don't get libcanberra and package kit errors at runtime. It's definitely not perfect, and it seems reasonable to assume that if the problems with linuxdeploy are ever fixed, these will be requirements, so I'm inclined to leave them in.

This is true; `canberra` does seem to load on RHEL distros. However, this is because installing the `gtk3` package on the host bring along the `libcanberra` package which includes `libcanberra.so.0`. So, when the `libcanberra-gtk-module.so` is loaded from within the AppImage, it needs `libcanberra.so.0`.....and despite it being in `usr/lib` of the AppImage, it ends up getting loaded from the host system. This is a bit precarious.....since there's no guarantee the `canberra` libraries in the AppImage will be compatible with those on the host. But I ultimately don't feel too strongly about this...
